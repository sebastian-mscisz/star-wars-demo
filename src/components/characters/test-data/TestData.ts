import { Characters, Character, Species } from 'src/common';

export const TestCharacter: Character = {
    name: 'Character1',
    url: 'https://swapi.info/api/people/1',
    homeworld: 'https://swapi.info/api/planets/1',
    species: ['species/1'],
    vehicles: ['https://swapi.info/api/vehicles/1'],
};

export const TestCharacters: Characters = [
    TestCharacter,
    {
        name: 'Character2',
        url: 'https://swapi.info/api/people/2',
        vehicles: ['https://swapi.info/api/vehicles/1', 'https://swapi.info/api/vehicles/2'],
        homeworld: 'https://swapi.info/api/planets/2',
        species: ['species/2'],
    },
    {
        name: 'Character3',
        url: 'https://swapi.info/api/people/3',
        vehicles: [],
        homeworld: 'https://swapi.info/api/planets/3',
        species: ['species/3'],
    },
    {
        name: 'Character4',
        url: 'https://swapi.info/api/people/4',
        vehicles: ['https://swapi.info/api/vehicles/3'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/4'],
    },
    {
        name: 'Character5',
        url: 'https://swapi.info/api/people/5',
        vehicles: ['https://swapi.info/api/vehicles/4', 'https://swapi.info/api/vehicles/5'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/5'],
    },
];

export const TestSpecies: Species = [
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species3',
        url: 'species/3',
    },
    {
        name: 'Species4',
        url: 'species/4',
    },
    {
        name: 'Species5',
        url: 'species/5',
    },
];
