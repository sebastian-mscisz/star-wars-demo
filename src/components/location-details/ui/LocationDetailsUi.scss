@import 'src/variables';

.sw-location-detail {
    font-size: 0.9rem;
    color: $base-color;
    background-color: $base-background-color;
    text-align: center;
    width: 80%;
    max-width: 550px;
    margin-top: 15px;
    border-radius: 10% 10% 10px 10px;
    border: 2px solid $base-color;
    display: flex;
    flex-direction: column;
    align-items: center;

    @media screen and (max-width: $small-bp) {
        margin-top: 25px;
    }

    &__name {
        font-size: 1rem;
    }

    &__image-wrapper {
        width: 95%;
        border-radius: 50%;
        overflow: hidden;
        transition: all 0.3s linear;
        outline: 2px solid $base-color;
        aspect-ratio: 1/1;
        background-color: $base-background-color;

        &__display-spinner {
            height: 100%;
            display: flex;
            justify-content: center;
        }

        &__hide-spinner {
            display: none;
        }

        &__hide-img {
            visibility: hidden;
        }

        & img {
            width: 100%;
            animation: rotating 60s linear infinite;
        }
    }

    &__details {
        text-align: left;
        width: 100%;
        line-height: 1.5;
        border-top: 2px solid $base-color;
        margin: 20px 0;

        & p {
            padding: 0 20px;
        }

        &__link {
            text-decoration: underline;
            font-weight: 700;
            color: $base-color;

            &:hover {
                color: $base-color-light;
            }
        }

        & hr {
            width: 100%;
        }
    }
}

@keyframes rotating {
    from {
        -ms-transform: rotate(0deg);
        -moz-transform: rotate(0deg);
        -webkit-transform: rotate(0deg);
        -o-transform: rotate(0deg);
        transform: rotate(0deg);
    }

    to {
        -ms-transform: rotate(360deg);
        -moz-transform: rotate(360deg);
        -webkit-transform: rotate(360deg);
        -o-transform: rotate(360deg);
        transform: rotate(360deg);
    }
}