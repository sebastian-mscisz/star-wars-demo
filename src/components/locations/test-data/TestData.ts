import { Locations, Location } from 'src/common';

export const TestLocation: Location = {
    name: 'Location1',
    url: 'https://swapi.info/api/planets/1',
    residents: ['https://swapi.info/api/people/1'],
    population: '1000',
};

export const TestLocations: Locations = [
    TestLocation,
    {
        name: 'Location2',
        url: 'https://swapi.info/api/planets/2',
        residents: ['https://swapi.info/api/people/2'],
        population: '5343',
    },
    {
        name: 'Location3',
        url: 'https://swapi.info/api/planets/3',
        residents: ['https://swapi.info/api/people/3'],
        population: '13333',
    },
    {
        name: 'Location4',
        url: 'https://swapi.info/api/planets/4',
        residents: ['https://swapi.info/api/people/4', 'https://swapi.info/api/people/5'],
        population: '9999',
    },
    {
        name: 'Location5',
        url: 'https://swapi.info/api/planets/5',
        residents: [],
        population: '999',
    },
];
