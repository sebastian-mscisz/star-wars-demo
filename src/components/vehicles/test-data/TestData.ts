import { Vehicles, Vehicle } from 'src/common';

export const TestVehicle: Vehicle = {
    name: 'Vehicle1',
    url: 'https://swapi.info/api/vehicles/1',
    vehicle_class: 'Air speeder',
    pilots: ['character/1', 'character/2'],
};

export const TestVehicles: Vehicles = [
    TestVehicle,
    {
        name: 'Vehicle2',
        url: 'https://swapi.info/api/vehicles/2',
        pilots: ['character/2'],
        vehicle_class: 'Land speeder',
    },
    {
        name: 'Vehicle3',
        url: 'https://swapi.info/api/vehicles/3',
        pilots: ['character/4'],
        vehicle_class: 'Water speeder',
    },
    {
        name: 'Vehicle4',
        url: 'https://swapi.info/api/vehicles/4',
        pilots: ['character/5'],
        vehicle_class: 'Fire speeder',
    },
    {
        name: 'Vehicle5',
        url: 'https://swapi.info/api/vehicles/5',
        pilots: ['character/5'],
        vehicle_class: 'Air speeder',
    },
];
