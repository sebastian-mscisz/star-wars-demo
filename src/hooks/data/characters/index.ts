export * from './useCharacters';
export * from './useCharacterDetails';
