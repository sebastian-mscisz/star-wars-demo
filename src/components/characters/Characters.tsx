import { withHandlers } from 'src/hoc';
import { CharactersUi } from './ui/CharactersUi';
import { useCharacters } from 'src/hooks';

const CharactersWithHandlers = withHandlers(CharactersUi);

export const Characters: React.FC = () => {
    const { data: characters, isLoading, loadingError } = useCharacters();

    return <CharactersWithHandlers isLoading={isLoading} characters={characters} hasError={!!loadingError} />;
};
