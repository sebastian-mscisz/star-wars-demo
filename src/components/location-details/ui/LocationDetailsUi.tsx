import React, { useState } from 'react';
import {
    Characters,
    LocationWithNeighbours,
    characterDetailsPagePath,
    getIdFromUrl,
    locationDetailsPagePath,
} from 'src/common';
import './LocationDetailsUi.scss';
import { NavLink } from 'react-router-dom';
import { PaginationButtons, Spinner } from 'src/components';

type LocationDetailsUiProps = {
    location?: LocationWithNeighbours;
    locationId: string;
    characters?: Characters;
};

export const LocationDetailsUi: React.FC<LocationDetailsUiProps> = ({ location, locationId, characters }) => {
    const [imgLoading, setImgLoading] = useState(true);

    if (!location) return null;

    const residents =
        characters
            ?.filter((character) => location.residents?.includes(character.url!))
            .map((el, i) => [
                i > 0 && ', ',
                <NavLink
                    key={`${el.name} ${i}`}
                    className="sw-location-detail__details__link"
                    to={characterDetailsPagePath(getIdFromUrl(el.url))}
                >
                    {el.name}
                </NavLink>,
            ]) || 'No famous residents';

    return (
        <>
            <div className="sw-location-detail">
                <div className="sw-location-detail__name">
                    <h2>{location.name}</h2>
                </div>
                <div className="sw-location-detail__image-wrapper">
                    <div className={`${imgLoading ? 'sw-location-detail__image-wrapper__display-spinner' : 'sw-location-detail__image-wrapper__hide-spinner'}`}>
                        <Spinner />
                    </div>
                    <img
                        className={`${imgLoading ? 'sw-location-detail__image-wrapper__hide_img' : ''}`}
                        src={`./locations/${locationId}.jpg`}
                        alt="Star wars location"
                        onLoad={() => setImgLoading(false)} />
                </div>
                <div className="sw-location-detail__details">
                    <p>
                        <strong>Population:</strong> {location.population}
                    </p>
                    <hr />
                    <p>
                        <strong>Residents:</strong> {residents?.length !== 0 ? residents : 'No famous residents'}
                    </p>
                </div>
            </div>
            <PaginationButtons
                leftPath={locationDetailsPagePath(getIdFromUrl(location.leftNeighbour.url))}
                rightPath={locationDetailsPagePath(getIdFromUrl(location.rightNeighbour.url))}
                center
            />
        </>
    );
};
