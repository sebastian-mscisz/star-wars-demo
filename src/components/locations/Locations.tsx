import { withHandlers } from 'src/hoc';
import { LocationsUi } from './ui/LocationsUi';
import { useLocations } from 'src/hooks';

const LocationsWithHandlers = withHandlers(LocationsUi);

export const Locations: React.FC = () => {
    const { data: locations, isLoading, loadingError } = useLocations();

    return <LocationsWithHandlers isLoading={isLoading} locations={locations} hasError={!!loadingError} />;
};
