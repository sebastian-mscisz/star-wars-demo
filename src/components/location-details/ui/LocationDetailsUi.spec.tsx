import { BrowserRouter } from 'react-router-dom';
import { LocationDetailsUi } from './LocationDetailsUi';
import { render } from '@testing-library/react';
import { TestCharacters } from '../../characters/test-data';
import { TestLocations } from '../../locations/test-data';

describe('LocationDetailsUi', () => {
    it('should render', () => {
        const { asFragment } = render(
            <LocationDetailsUi
                location={{ ...TestLocations[1], leftNeighbour: TestLocations[0], rightNeighbour: TestLocations[2] }}
                characters={TestCharacters}
                locationId={'3'}
            />,
            { wrapper: BrowserRouter },
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
