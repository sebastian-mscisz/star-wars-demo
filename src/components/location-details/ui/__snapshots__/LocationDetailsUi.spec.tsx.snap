// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationDetailsUi should render 1`] = `
<DocumentFragment>
  <div
    class="sw-location-detail"
  >
    <div
      class="sw-location-detail__name"
    >
      <h2>
        Location2
      </h2>
    </div>
    <div
      class="sw-location-detail__image-wrapper"
    >
      <div
        class="sw-location-detail__image-wrapper__display-spinner"
      >
        <div
          class="sw-spinner"
        >
          <span
            style="display: inherit; position: relative; width: 25px; height: 5px; overflow: hidden; background-color: black; background-clip: padding-box; border-radius: 5px;"
          >
            <span
              style="position: absolute; height: 5px; overflow: hidden; background-color: rgb(0, 0, 0); background-clip: padding-box; display: block; border-radius: 2px; will-change: left, right; animation-fill-mode: forwards; animation: react-spinners-BarLoader-long Infinitys  cubic-bezier(0.65, 0.815, 0.735, 0.395) infinite;"
            />
            <span
              style="position: absolute; height: 5px; overflow: hidden; background-color: rgb(0, 0, 0); background-clip: padding-box; display: block; border-radius: 2px; will-change: left, right; animation-fill-mode: forwards; animation: react-spinners-BarLoader-short Infinitys Infinitys cubic-bezier(0.165, 0.84, 0.44, 1) infinite;"
            />
          </span>
          <span
            style="display: inherit; position: relative; width: 100px; height: 5px; overflow: hidden; background-color: rgba(0, 0, 0, 0.2); background-clip: padding-box; border-radius: 5px;"
          >
            <span
              style="position: absolute; height: 5px; overflow: hidden; background-color: rgb(0, 0, 0); background-clip: padding-box; display: block; border-radius: 2px; will-change: left, right; animation-fill-mode: forwards; animation: react-spinners-BarLoader-long 2.1s  cubic-bezier(0.65, 0.815, 0.735, 0.395) infinite;"
            />
            <span
              style="position: absolute; height: 5px; overflow: hidden; background-color: rgb(0, 0, 0); background-clip: padding-box; display: block; border-radius: 2px; will-change: left, right; animation-fill-mode: forwards; animation: react-spinners-BarLoader-short 2.1s 1.15s cubic-bezier(0.165, 0.84, 0.44, 1) infinite;"
            />
          </span>
        </div>
      </div>
      <img
        alt="Star wars location"
        class="sw-location-detail__image-wrapper__hide_img"
        src="./locations/3.jpg"
      />
    </div>
    <div
      class="sw-location-detail__details"
    >
      <p>
        <strong>
          Population:
        </strong>
         5343
      </p>
      <hr />
      <p>
        <strong>
          Residents:
        </strong>
         
        <a
          class="sw-location-detail__details__link"
          href="/character/2"
        >
          Character2
        </a>
      </p>
    </div>
  </div>
  <div
    class="sw-pagination-section sw-pagination-section--center"
    data-testid="pagination-section"
  >
    <div
      class="sw-pagination-button "
      data-testid="left-pagination-button"
    >
      ←
    </div>
    <div
      class="sw-pagination-button "
      data-testid="right-pagination-button"
    >
      →
    </div>
  </div>
</DocumentFragment>
`;
